s2e.so: lazily initialize function models

adrianherrera opened the pull request:
s2e.so may segfault when using LD_LIBRARY_PATH because modelled functions may be called before `__libc_start_main` is called. In this case all of the `orig_*` function pointers will be `NULL`, causing a segfault.

adrianherrera committed:
s2e.so: initialize function models on demand

s2e.so may segfault when using LD_LIBRARY_PATH because modelled functions may
be called before __libc_start_main is called. In this case all of the orig_*
function pointers will be NULL, causing a segfault.

vitalych commented on linux/function_models/libz_wrapper.c:
```diff
@@ -31,6 +31,10 @@
 
 uint32_t crc32(uint32_t crc, const uint8_t *buf, unsigned len) {
     if (!g_enable_function_models) {
+        if (!orig_crc32) {
```
Would it be possible to turn these functions into a template? It would reduce the amount of code duplication.
We'd have to switch this file to c++, I hope that would work.

adrianherrera replied:
Sure, I'll look into it.

vitalych reviewed (changes requested):
.

adrianherrera committed:
linux: restructured common/include

Moved s2e_so.h and function_models/models.h so that they are local only to
Linux guests (they are not used elsewhere).

Also made test_case_generator consistent with the other "command" header files.

adrianherrera committed:
linux: cleaned up build warnings

adrianherrera commented:
I tried changing to C++ templates but then the LD_PRELOAD stopped working (`S2E_SYM_ARGS` stopped working) and I didn't want to debug, so I reverted my changes.

I'm happy to implement some macros if you want?

vitalych commented:
macros  would work too

vitalych commented on common/include/s2e/function_models/models.h:
```diff
@@ -36,6 +36,9 @@
 
+//
+// Types for modelled functions
+
```
Missing //

vitalych commented:
I suppose switching to c++ causes it to link with the c++ library, which messes things up

adrianherrera committed:
function_models: refactored common code as macro

Reduce code duplication.

adrianherrera commented:
Ok I've refactored with macros.

Ready for re-review.

vitalych commented:
There seems to be a code style issue: http://jenkins.s2e.systems/job/s2e-clang-format/16/console
Not sure why it didn't get reported to github, I'll have to debug.

adrianherrera commented:
Ah ok, cool! I've fixed the issue anyway.
